tuxemon/core/states/world/worldstate.py

```diff
@@ -598,6 +604,30 @@ def get_explicit_tile_exits(self, position, tile, skip_nodes):
         except KeyError:
             pass
 
+    def tile_blocks(self, entity, tile):
+        """ Verifies if a collision tile is supposed to block a given type of NPC
+
+        Three checks are preformed:
+        If the NPC is a player, return True if "players" exists in the blocks
+        If the NPC isn't a player, return True if "npcs" exists in the blocks
+        If the slug of the NPC is present in the blocks, return True
+
+        :param entity: the player or NPC being checked
+        :param tile: the data of the tile being checked
+
+        :rtype: bool
+        """
+        if entity and tile and 'blocks' in tile:
+            for block in tile['blocks'].split():
+                if block == 'players' and entity.isplayer:
+                    return True
+                if block == 'npcs' and not entity.isplayer:
+                    return True
+                if block == entity.slug:
+                    return True
+            return False
+        return True
+
     def get_exits(self, position, collision_map=None, skip_nodes=None):
```

Comment:
This is going to break pathfinding.  Consider the situation where the player (or different NPC) is standing in a position which conditionally blocks the NPC in a different way than the one pathfinding (or vice-versa).  When the pathfinding code runs, the path will be different depending on what npc is in that spot with the blocking rules.  So a path that might be perfectly valid for one npc will not be found because while looking for paths, the code uses blocking rules that apply for a different NPC.

Reply:
Doesn't the pathfinding code use the `get_exits` function where the check is implemented to get its rules? So if a path is invalid for a specific NPC, the pathfinding code would know and go for a different path instead, accounting only the collisions that affect that NPC?

Need to wrap my mind around around exactly how this is the case first. If it's so I'll think of what can be done about it, this would indeed be a problem.

Reply:
no.  get_exists only knows about the current state of the map.  with these changes get_exits changes depending on what NPC is in a tile position.  since all NPCs will pathfind using this method, they will now be using the rules of the tiles where some NPC is sitting.

Reply:
I see. Is there an easy way to fix this? If not I could go with a simple solution for now: Adding a TODO note and in the meantime, disable this filter entirely for pathfinding. IIRC there is a `npc.pathfinding` boolean which is True when the NPC is navigating that way. This might however be a bad idea, since if the NPC stops pathfinding inside a collision it might get stuck there forever.

Reply:
no hacks, please.  i don't have an easy solution right now.  essentially, all npc movement and pathfinding need to carry "context" with them because the exits change depending on who is looking that them.

Reply:
i would accept disabling teleports for npcs and an alternative

Reply:
Ok. I'll need to wait for you or another dev to provide a solution to this one: It delves into more difficult aspects that I'm not sure how to approach.

One clue, even if this is probably obvious: The new `def tile_blocks(self, entity, tile)` function can be reused anywhere, given we have access to the NPC entity we're checking and the tile data of the spot we're moving to. If there's a specific point in the pathfinding code where the same checks would be needed, it can probably be called there as well.